src/CircuitOptimizer.cpp

```diff
@@ -667,4 +667,79 @@ namespace qc {
     void CircuitOptimizer::decomposeTeleport([[maybe_unused]] QuantumComputation& qc) {
     }
 
+    void CircuitOptimizer::eliminateResets(QuantumComputation& qc) {
+        /// TODO: substitute reset operations with new qubits and shift subsequent operations
+
+        //      ┌───┐┌─┐     ┌───┐┌─┐            ┌───┐┌─┐ ░
+        // q_0: ┤ H ├┤M├─|0>─┤ H ├┤M├       q_0: ┤ H ├┤M├─░─────────
+        //      └───┘└╥┘     └───┘└╥┘   -->      └───┘└╥┘ ░ ┌───┐┌─┐
+        // c: 2/══════╩════════════╩═       q_1: ──────╫──░─┤ H ├┤M├
+        //            0            1                   ║  ░ └───┘└╥┘
+        //                                  c: 2/══════╩══════════╩═
+        //                                             0          1
+        auto replacementMap = std::map<dd::Qubit, dd::Qubit>();
+        for (auto it = qc.ops.begin(); it != qc.ops.end(); it++) {
+            if (!replacementMap.empty()) {
+                if ((*it)->isStandardOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isNonUnitaryOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isCompoundOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+                if ((*it)->isClassicControlledOperation()) {
+                    changeTargets(reinterpret_cast<Operation&>(*it), replacementMap);
+                    changeControls(reinterpret_cast<Operation&>(*it), replacementMap);
+                }
+            }
+            if ((*it)->getType() == qc::Reset) {
+                auto indexAddQubit = static_cast<dd::Qubit>(qc.getNqubits());
+                qc.addQubit(indexAddQubit, indexAddQubit, indexAddQubit);
+                for (const auto& target: (*it)->getTargets()) {
+                    replacementMap.insert(std::pair(static_cast<dd::Qubit>(target), static_cast<dd::Qubit>(indexAddQubit)));
+                }
+                it = qc.erase(it);
+            }
+        }
+    }
+
+    void CircuitOptimizer::changeTargets(Operation& op, std::map<dd::Qubit, dd::Qubit> replacementMap) {
+        auto targetIt = op.getTargets().begin();
+        while (targetIt != op.getTargets().end()) {
+            auto newTarget = replacementMap.find((static_cast<dd::Qubit>(*targetIt)));
+            if (newTarget != replacementMap.end()) {
+                op.setTargets(static_cast<Targets>(newTarget->second));
+            }
+            targetIt++;
+        }
+    }
```

Comment:
As described above, this method should just receive a reference to a vector of targets (and a `const` reference to the replacement map) and modify the vector accordingly. See below for a suggestion:
```suggestion
    void CircuitOptimizer::changeTargets(Targets& targets, const std::map<dd::Qubit, dd::Qubit>& replacementMap) {
        for(auto& target: targets) {
            auto newTargetIt = replacementMap.find(target);
            if (newTargetIt != replacementMap.end()) {
                target = newTarget->second;
            }
        }
    }
```